File: python/mod/api/entityModule.py
from .jni import findJavaCls, CAST_TYPE, floatSplit, numberTupleJoin, ClientOnly
from json import dumps
lambda: "By Zero123"

_ENTITY_MODULE = "org/zero123/PyScriptEngine/ModSdk/EntityModule"
_PROJECTILE_MODULE = "org/zero123/PyScriptEngine/ModSdk/ProjectileModule"
_CMD_MODULE = "org/zero123/PyScriptEngine/ModSdk/Command"

def _serverGetEntityPos(entityId: str):
    """ 获取实体位置 """
    return floatSplit(findJavaCls(_ENTITY_MODULE, "_serverGetEntityPos", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    ))

def _clientGetEntityPos(entityId: str):
    """ 客户端获取实体位置 """
    return floatSplit(findJavaCls(_ENTITY_MODULE, "_clientGetEntityPos", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    ))

def _serverSetEntityPos(entityId: str, pos: tuple):
    """ 设置实体位置 """
    return findJavaCls(_ENTITY_MODULE, "_serverSetEntityPos", [CAST_TYPE.STRING, CAST_TYPE.STRING], CAST_TYPE.VOID).call(
        entityId, numberTupleJoin((pos[0], pos[1], pos[2]))
    )

def _serverGetEntityRot(entityId: str):
    """ 获取实体旋转欧拉角 """
    return floatSplit(findJavaCls(_ENTITY_MODULE, "_serverGetEntityRot", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    ))

def _clientGetEntityRot(entityId: str):
    return floatSplit(findJavaCls(_ENTITY_MODULE, "_clientGetEntityRot", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    ))

def _serverShootProjectile(entityId: str, entityIdentifier: str, dataMap: dict) -> str:
    """ 创建并发射投掷物 """
    return findJavaCls(_PROJECTILE_MODULE, "_serverShootProjectile", [CAST_TYPE.STRING, CAST_TYPE.STRING, CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId, entityIdentifier, dumps(dataMap)
    )

def _serverGetEntityTypeName(entityId: str) -> str:
    """ 获取实体类型名称(标识符) """
    return findJavaCls(_ENTITY_MODULE, "_serverGetEntityTypeName", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    )

def _clientGetEntityTypeName(entityId: str) -> str:
    """ 客户端获取实体类型名称(标识符) """
    return findJavaCls(_ENTITY_MODULE, "_clientGetEntityTypeName", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    )

def _serverCheckIsPlayer(entityId: str) -> bool:
    """ 检查实体是否为玩家 """
    return bool(findJavaCls(_ENTITY_MODULE, "_serverCheckIsPlayer", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

def _clientCheckIsPlayer(entityId: str) -> bool:
    """ 客户端检查实体是否为玩家 """
    return bool(findJavaCls(_ENTITY_MODULE, "_clientCheckIsPlayer", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

def _serverCheckEntityAlive(entityId: str) -> bool:
    """ 检查实体是否存活 """
    return bool(findJavaCls(_ENTITY_MODULE, "_serverCheckEntityAlive", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

def _clientCheckEntityAlive(entityId: str) -> bool:
    """ 客户端检查实体是否存活 """
    return bool(findJavaCls(_ENTITY_MODULE, "_clientCheckEntityAlive", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

def _serverGetWorldEntityList() -> list:
    """ 获取世界实体id列表 """
    data = str(findJavaCls(_ENTITY_MODULE, "_serverGetWorldEntityList", [], CAST_TYPE.STRING).call())
    if not data:
        return []
    return list(data.split(" "))

def _clientGetWorldEntityList() -> list:
    """ 客户端获取世界实体id列表 """
    data = str(findJavaCls(_ENTITY_MODULE, "_clientGetWorldEntityList", [], CAST_TYPE.STRING).call())
    if not data:
        return []
    return list(data.split(" "))

@ClientOnly
def _clientGetLocalPlayerId() -> str:
    """ 客户端获取本地玩家ID """
    return findJavaCls(_ENTITY_MODULE, "_clientGetLocalPlayerId", [], CAST_TYPE.STRING).call()

def _serverKillEntity(entityId: str) -> bool:
    """ 杀死实体 """
    return bool(findJavaCls(_ENTITY_MODULE, "_serverKillEntity", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

def _serverDestroyEntity(entityId: str) -> bool:
    """ 销毁实体 """
    return bool(findJavaCls(_ENTITY_MODULE, "_serverDestroyEntity", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

def _serverGetAllPlayerId() -> list:
    """ 获取所有玩家ID """
    data = str(findJavaCls(_ENTITY_MODULE, "_serverGetAllPlayerId", [], CAST_TYPE.STRING).call())
    if not data:
        return []
    return list(data.split(" "))

def _serverGetEntityTargetId(entityId: str) -> str:
    """ 获取实体目标ID """
    return findJavaCls(_ENTITY_MODULE, "_serverGetEntityTargetId", [CAST_TYPE.STRING], CAST_TYPE.STRING).call(
        entityId
    )

def _serverSendMessage(playerId: str, message: str):
    """ 向指定玩家发送消息 """
    return findJavaCls(_ENTITY_MODULE, "_serverSendMessage", [CAST_TYPE.STRING, CAST_TYPE.STRING], CAST_TYPE.VOID).call(
        playerId, message
    )

def _serverSendMessageAll(message: str):
    """ 向所有玩家发送消息 """
    return findJavaCls(_ENTITY_MODULE, "_serverSendMessageAll", [CAST_TYPE.STRING], CAST_TYPE.VOID).call(
        message
    )

def _clientSendMessage(message: str):
    """ 向客户端发送消息 """
    return findJavaCls(_ENTITY_MODULE, "_clientSendMessage", [CAST_TYPE.STRING], CAST_TYPE.VOID).call(
        message
    )

def _setCommand(cmdStr: str, entityId: str="", showOutput: bool=False) -> bool:
    """ 设置执行命令 """
    return bool(findJavaCls(_CMD_MODULE, "_setCommand", [CAST_TYPE.STRING, CAST_TYPE.STRING, CAST_TYPE.INT], CAST_TYPE.INT).call(
        cmdStr, entityId, int(showOutput)
    ))

def _serverGetEntityDmId(entityId: str) -> int:
    """ 获取实体所在维度ID, 异常返回-1, 原版维度返回0-2, 三方JE自定义维度返回其他映射负数值(仅运行时临时分配) """
    return int(findJavaCls(_ENTITY_MODULE, "_serverGetEntityDmId", [CAST_TYPE.STRING], CAST_TYPE.INT).call(
        entityId
    ))

File: java/src/main/java/org/zero123/PyScriptEngine/ModSdk/EntityModule.java
package org.zero123.PyScriptEngine.ModSdk;

import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.network.chat.Component;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.Vec3;
import net.neoforged.neoforge.server.ServerLifecycleHooks;
import org.zero123.PyScriptEngine.Utils.EntityUtil;
import org.zero123.PyScriptEngine.Utils.WorldUtil;

import java.util.ArrayList;

public class EntityModule
{
    public static class EntityRot
    {
        public float x;
        public float y;
        public EntityRot(float x, float y)
        {
            this.x = x;
            this.y = y;
        }
    }

    // 获取实体的rot旋转角
    public static EntityRot getEntityRot(Entity entity)
    {
        return new EntityRot(entity.getXRot(), entity.getYRot());
    }

    // 基于rot转换为向量
    public static Vec3 toDirectionVector(float yaw, float pitch)
    {
        double radYaw = Math.toRadians(yaw);
        double radPitch = Math.toRadians(pitch);
        double x = -Math.cos(radPitch) * Math.sin(radYaw);
        double y = -Math.sin(radPitch);
        double z = Math.cos(radPitch) * Math.cos(radYaw);
        return new Vec3(x, y, z);
    }

    public static Vec3 toDirectionVector(EntityRot rot)
    {
        return toDirectionVector(rot.x, rot.y);
    }

    // ====================== 绑定函数 ======================

    // 获取实体坐标
    public static String _serverGetEntityPos(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isEmpty())
        {
            return "";
        }
        final var pos = entityOpt.get().position();
        Double[] posArray = new Double[] {
            pos.x, pos.y, pos.z
        };
        return Utils.arrayToString(posArray);
    }

    // 客户端获取实体坐标
    public static String _clientGetEntityPos(String entityId)
    {
        var entityOpt = EntityUtil.clientGetEntityByUUID(entityId);
        if(entityOpt.isEmpty())
        {
            return "";
        }
        final var pos = entityOpt.get().position();
        Double[] posArray = new Double[] {
                pos.x, pos.y, pos.z
        };
        return Utils.arrayToString(posArray);
    }

    // 设置实体坐标
    public static void _serverSetEntityPos(String entityId, String posBytes)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            var posArray = Utils.parseDoubleArray(posBytes);
            final var entity = entityOpt.get();
            entity.teleportTo(posArray[0], posArray[1], posArray[2]);
            entity.setDeltaMovement(Vec3.ZERO);
/*            if (entity.get() instanceof ServerPlayer player)
            {
                // 玩家同步并刷新客户端
                player.teleportTo(posArray[0], posArray[1], posArray[2]);
            }*/
        }
    }

    // 获取实体旋转角
    public static String _serverGetEntityRot(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            final var entity = entityOpt.get();
            return Utils.arrayToString(new Float[] { entity.getXRot(), entity.getYRot() });
        }
        return "";
    }

    // 客户端获取实体旋转角
    public static String _clientGetEntityRot(String entityId)
    {
        var entityOpt = EntityUtil.clientGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            final var entity = entityOpt.get();
            return Utils.arrayToString(new Float[] { entity.getXRot(), entity.getYRot() });
        }
        return "";
    }

    // 获取实体的标识符
    public static String _serverGetEntityTypeName(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            return EntityType.getKey(entityOpt.get().getType()).toString();
        }
        return "";
    }

    // 客户端获取实体标识符
    public static String _clientGetEntityTypeName(String entityId)
    {
        var entityOpt = EntityUtil.clientGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            return EntityType.getKey(entityOpt.get().getType()).toString();
        }
        return "";
    }

    // 判断实体类型是不是玩家
    public static int _serverCheckIsPlayer(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            if(entityOpt.get() instanceof Player)
            {
                return 1;
            }
        }
        return 0;
    }

    // 客户端判断实体是不是玩家
    public static int _clientCheckIsPlayer(String entityId)
    {
        var entityOpt = EntityUtil.clientGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            if(entityOpt.get() instanceof Player)
            {
                return 1;
            }
        }
        return 0;
    }

    // 判断实体是否存在
    public static int _serverCheckEntityAlive(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        return entityOpt.isPresent() ? 1 : 0;
    }

    // 客户端判断实体是否存在
    public static int _clientCheckEntityAlive(String entityId)
    {
        var entityOpt = EntityUtil.clientGetEntityByUUID(entityId);
        return entityOpt.isPresent() ? 1 : 0;
    }

    // 【服务端】获取世界实体列表
    public static String _serverGetWorldEntityList()
    {
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        ArrayList<String> entityIds = new ArrayList<>();
        if (server != null)
        {
            for (ServerLevel level : server.getAllLevels())
            {
                // 遍历当前维度所有实体
                for (Entity entity : level.getEntities().getAll())
                {
                    entityIds.add(entity.getUUID().toString());
                }
            }
        }
        return Utils.joinWithSpace(entityIds);
    }

    // 【客户端】获取加载中的所有实体列表
    public static String _clientGetWorldEntityList()
    {
        final var clientLevel = Minecraft.getInstance().level;
        ArrayList<String> entityIds = new ArrayList<>();
        if (clientLevel != null)
        {
            // 遍历当前维度所有实体
            for (Entity entity : clientLevel.entitiesForRendering())
            {
                entityIds.add(entity.getUUID().toString());
            }
        }
        return Utils.joinWithSpace(entityIds);
    }

    // 【客户端】获取本地玩家ID
    public static String _clientGetLocalPlayerId()
    {
        final var player = Minecraft.getInstance().player;
        if (player != null)
        {
            return player.getUUID().toString();
        }
        return "";
    }

    // 杀死特定实体
    public static int _serverKillEntity(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            final var entity = entityOpt.get();
            if (entity.level() instanceof ServerLevel serverLevel)
            {
                entity.kill(serverLevel);
                return 1;
            }
            return 0;
        }
        return 0;
    }

    // 销毁特定实体
    public static int _serverDestroyEntity(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isPresent())
        {
            final var entity = entityOpt.get();
            if (entity.level() instanceof ServerLevel && !(entity instanceof Player))
            {
                entity.discard();
                return 1;
            }
            return 0;
        }
        return 0;
    }

    // 获取所有玩家ID
    public static String _serverGetAllPlayerId()
    {
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        if (server == null)
        {
            return "";
        }

        StringBuilder sb = new StringBuilder();
        for (ServerPlayer player : server.getPlayerList().getPlayers())
        {
            sb.append(player.getUUID()).append(" ");
        }
        if (!sb.isEmpty())
        {
            sb.setLength(sb.length() - 1); // 去掉最后一个空格
        }
        return sb.toString();
    }

    // 获取指定实体的目标id 如果不存在返回空字符串
    public static String _serverGetEntityTargetId(String entityId)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isEmpty())
        {
            return "";
        }
        final var entity = entityOpt.get();
        if (entity instanceof Mob mob)
        {
            LivingEntity target = mob.getTarget();
            if (target != null)
            {
                return target.getUUID().toString();
            }
        }
        return "";
    }

    /** 向指定玩家发送一条服务端消息（系统消息） **/
    public static void _serverSendMessage(String entityId, String message)
    {
        var entityOpt = EntityUtil.serverGetEntityByUUID(entityId);
        if(entityOpt.isEmpty())
        {
            return;
        }
        if(entityOpt.get() instanceof ServerPlayer player)
        {
            player.sendSystemMessage(Component.literal(message));
        }
    }

    /** 向所有玩家发送消息 **/
    public static void _serverSendMessageAll(String message)
    {
        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
        if (server == null)
        {
            return;
        }
        Component text = Component.literal(message);
        for (ServerPlayer player : server.getPlayerList().getPlayers())
        {
            player.sendSystemMessage(text);
        }
    }

    /** 客户端玩家创建系统消息 **/
    public static void _clientSendMessage(String msg)
    {
        Minecraft mc = Minecraft.getInstance();
        LocalPlayer player = mc.player;
        if (player != null)
        {
            player.displayClientMessage(Component.literal(msg), false);
        }
    }

    // 服务端获取实体所在维度
    public static int _serverGetEntityDmId(String entityId)
    {
        return WorldUtil.serverGetEntityDimensionId(EntityUtil.serverGetEntityByUUID(entityId).orElse(null));
    }
}
